keps/sig-network/5311-relaxed-validation-for-service-names/README.md

```diff
@@ -0,0 +1,761 @@
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user, I want to name my Deployment "12345", and I'd like a Service and Ingress object to match that name.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+1. Services are responsible for creating DNS records (ie: `<service-name>.<namespace>.svc.cluster.local>`). To confirm that downstream systems will support the new validation, we will conduct compatibility testing by:
+   - Verifying that DNS providers used in Kubernetes clusters can handle the new service name format.
+   - Running integration tests to ensure that dependent components such as Ingress controllers and service discovery mechanisms function correctly with the updated validation.
+2. The Ingress resource references Service and will also need a relaxed validation on its reference to the Service
+3. Downstream applications may perform validation on fields relating to Service names (ie: [ingress-ngnix](https://github.com/kubernetes/ingress-nginx/blob/d3ab5efd54f38f2b7c961024553b0ad060e2e916/internal/ingress/annotations/parser/validators.go#L190-L197)
+
+## Design Details
+
+Introduce a new feature gate named RelaxedServiceNameValidation is disabled by default.
+
+When enabled, the feature gate will use the [NameIsDNSSubdomain](https://github.com/kubernetes/kubernetes/blob/3196c9946355c1d20086f66c22e9e5364fb0a56f/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L37-L42) validation for new Services.
+
+Since the relaxed check allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+
+When the feature gate is disabled:
+
+- Creation of Services will use the previous RFC-1035 validation for `.metadata.name`
+- Creation of Ingress will use the previous RFC-1035 validation for `.spec.[]rules.http.[]paths.backend.service.name`
+- Updates of Services will use the new RFC-1123 validation for `.metadata.name`
+- Updates of Ingress will use the previous RFC-1035 validation for `.spec.[]rules.http.[]paths.backend.service.name`
```

Comment:
So, if upgrade and have an Ingress object with a reference of sarvid valid for RFC-1123, and then downgrade, any update in this Ingress object will fail?

Reply:
Whoops, that's a typo. Thanks for catching it. It's supposed to be RFC-1123